### PR TITLE
Refactor build_data

### PR DESCRIPTION
**Pull request by jbcden**
This is a first pass at addressing https://github.com/steveklabnik/rustdoc/issues/55.

### CODE REVIEW

**Comment by jbcden** [comment]
@steveklabnik Here's a first pass at refactoring the `build_data` function. I'm not sure if this is what you were thinking but let me know your thoughts!

File: src/lib.rs [hunk 1]
```diff
@@ -321,10 +317,5 @@ struct Crate {
     id: analysis::Id,
     name: String,
     docs: String,
-    modules: Vec<String>,
-}
-
-#[derive(Debug)]
-enum Item {
-    Module { name: String, docs: String },
+    metadata: Vec<Metadata>,
```
Comment (steveklabnik) [inline src/lib.rs]: 👍

File: src/lib.rs [hunk 2]
```diff
@@ -190,21 +192,15 @@ impl DocData {
...
     }
 
-    fn build_data(
-        config: &Config,
-        root_id: analysis::Id,
-        krate: &mut Crate,
-    ) -> Result<HashMap<String, Item>> {
-        let mut data = HashMap::new();
-
+    fn build_data(config: &Config, root_id: analysis::Id, krate: &mut Crate) -> Result<()> {
```
Comment (steveklabnik) [inline src/lib.rs]: So, this still ends up leaving this function and taking a `&mut Crate`; what do you think about possibly inlining this into the constructor? does it being its own function make sense still?

Comment (jbcden) [inline src/lib.rs]: @steveklabnik I go back and forth on that. As I said before, the main reason passing the crate in felt weird to me was that that seemed to signal that this function's main focus should be on the crate. Now that it is all about the crate, I don't feel this is too weird, though the name should definitely be changed to reflect this new role too.

Comment (steveklabnik) [inline src/lib.rs]: yeah let's just inline it for now 👍

File: src/lib.rs [hunk 3]
```diff
@@ -213,14 +209,11 @@ impl DocData {
         for def in defs.iter() {
```
Comment (mgattozzi) [inline src/lib.rs]: I think it might make more sense to turn this from `iter` to `into_iter` that way the clones below can be avoided.

Comment (mgattozzi) [inline src/lib.rs]: The reason being is defs isn't used after this loop at all so consuming the values should be fine.

Comment (steveklabnik) [inline src/lib.rs]: yeah i was certainly cargo-culting this code; getting rid of eventual clones would be great!

**Comment by jbcden** [comment]
> yeah let's just inline it for now 👍

@steveklabnik I inlined it in f91ddad.

**Comment by jbcden** [comment]
@mgattozzi Thanks for the advice! I'm still pretty new to Rust so I'm sure I miss a lot of stuff like that. I switched to `into_iter` and removed the clones in f91ddad and d27a90e.

**Comment by mgattozzi** [comment]
Woo awesome! :D

**Pull request merged**